### DISCUSSION TITLE
Adding a banner for non-recent documentation versions

### DISCUSSION THREAD

**Discussion by JulioV** [squidfunk/mkdocs-material#2084] [Ideas]
- [ x] I've read the [contribution guidelines][1] and agree with them

__I want to suggest an idea and checked that ...__

- [ x] ... to my best knowledge, my idea wouldn't break something for other users
- [ x] ... the documentation does not mention anything about my idea
- [ x] ... there are no open or closed issues that are related to my idea

## Description
Create a banner at the top of each page on old versions of the docs (when managed by `mike`) that point to the latest (most recent) version

### Use Cases

Whenever a user lands on an old docs page, they can be redirected to the most recent version like `readthedocs`

### Screenshots / Mockups

There are some suggestions in this issue https://github.com/jimporter/mike/issues/26 (see screenshot below)

There could be a flag to activate this feature  on every version (and its aliases) except for the most recent one based on `mike`'s `versions.json` indexes:

```
[
  {"version": "1.0", "title": "1.0.1", "aliases": ["latest"]},
  {"version": "0.9", "title": "0.9", "aliases": []}
]
```

The actual content and look and feel could be customizable like the announcement section of the theme

<img width="1625" alt="image" src="https://user-images.githubusercontent.com/5264981/101108655-02ed9c80-35a3-11eb-99c1-ee5823c254f6.png">

  [1]: https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md

**Comment by squidfunk** [comment]
Thanks for suggesting! I agree that this is a good idea which we should investigate. However, from the top of my head, there may be problems with content shift due to the fact that the `mike` integration is all-JavaScript, which means that the version selector is rendered after the initial load. Thus, the notice will be displayed _after_ the initial render, shifting the content downwards, which is penalized by reports like Lighthouse and in general is not a good idea.

Material for MkDocs painstakingly tries to minimize content shift and it does a very good job at that, so a possible solution for displaying a banner should not introduce this problem.

**Comment by squidfunk** [comment]
I converted this issue to a discussion, as I'd like to get some feedback before we decide whether and how we should implement it. Also note that there are problems with content shift that would need to be addressed somehow, as I stated in my last comment.

**Comment by miguno** [comment]
Adding my +1 to this feature suggestion.

> Material for MkDocs painstakingly tries to minimize content shift and it does a very good job at that, so a possible solution for displaying a banner should not introduce this problem.

This would be great indeed for UX.

That said, in my case it's much more important to notify users that they are looking at an old version of the content, even if this required content shifting for older versions. As I understand, no shift would happen for the 'latest' version this shift shouldn't happen, and (in my case) the great majority of traffic is for the latest version.

**Comment by JulioV** [comment]
I agree with @miguno, a small content shift is preferable to users browsing the wrong documentation version.

Is it possible to achieve no shift? One approach would be to have `mike` override the `announce` block on old versions (folders) during a `mike deploy --push` but I don't think is something they would want to integrate.

A popup would be another alternative but people dislike those more than content shifts I'd guess.

**Reply by squidfunk** [reply]
Besides the content shift, there are other problems with this. Where would the content come from? The problem is that versions are injected via JavaScript (as `mike` generates a `versions.json` which is used by the selector), which means we're in JavaScript land and post-build. I may revisit this at some point but for now I don't see a clear path to solve it generically.

**Reply by miguno** [reply]
One option is to always show the banner 'space' (i.e., make room for it on a page), but only display the warning _text_ when old content is being browsed. This prevents content shifting because the banner space is always there.

Of course, this isn't an option when the web design requires that there must be no such 'wiggle room' aka extra space.

**Reply by squidfunk** [reply]
Where does the warning text come from? How is it localized?

**Reply by miguno** [reply]
I don't know enough about MkDocs/Material yet to give a confident answer, but wouldn't this be similar to where other content (incl. localization) is coming from? Because this would be pre-build. The `mike` related setup, which is post-build and using JavaScript/CSS, would only be responsible for hiding (when viewing 'latest' content) vs. unhiding (when viewing old content) the pre-built warning.

My comment above was specifically about addressing the content shift issue, where the idea was to essentially always shift the content, but in the case of reading 'the latest' docs we'd trick the reader into not being aware of this because there's no text in the pre-shifted space.

**Reply by JulioV** [reply]
I agree the warning would be prebuilt, perhaps the user can override the link to the latest version (to account for specific aliases or external links), and JS would be in charge of showing it for old versions since the text span would be hidden by default (not the div container) in case JS is disabled

**Reply by JulioV** [reply]
Doing a naive inspection of the HTML layout, `class="md-main__inner md-grid"` already has a top margin of 1.5rem, so adding an always visible div of approximately that height should not change much how pages look at the moment. The text span containing the built-in warning would be the one being revealed by JS for old versions.

**Reply by squidfunk** [reply]
Okay, thanks everybody for the input! That might work. I'm unsure about positioning, but in general:

1. The text would be pulled from the translations, thus it would be generic and tied to the configured language. In case the authors want to change it, this is possible via [template overrides](https://squidfunk.github.io/mkdocs-material/setup/changing-the-language/#custom-translations)
2. The banner will be rendered with JavaScript when the version selector is rendered and a non-latest version is detected. The version alias for the latest version will be `latest`. If another version is active, the notice is rendered
3. What about rendering the notice sticky at the bottom of the screen? This way we would not have any content shift.

**Reply by JulioV** [reply]
1 and 2 sound good.

Not sure about 3, people tend to completely ignore things at the bottom, especially if the page is long. Could we pre-render a `div` block at the top containing the hidden warning `span` only when `extra:  version:  method: mike` and un-hide the warning `span` when the version alias is not `latest`?

**Reply by squidfunk** [reply]
Regarding 3, it would be fixed to the viewport, not the bottom of the document. However, CSS can be overridden and the banner can be positioned as desired by the author. Material for MkDocs would only provide a sane default.

**Reply by JulioV** [reply]
ah got it, that makes sense and works for me

**Reply by miguno** [reply]
Yes to 1 and 2.

Comment for 2:
> The banner will be rendered with JavaScript when the version selector is rendered and a non-latest version is detected. The version alias for the latest version will be latest. If another version is active, the notice is rendered

`mike` lets you set a default version with:

```
$ mike set-default [version-or-alias]
```
If possible, this should be honored. If not possible, then 'latest' seems like a good default to me.

Yes to 3 (viewport, not bottom of the document).

**Reply by squidfunk** [reply]
Heads-up: I'm close to finishing a solution with the following properties:

1. Banner is rendered at the top, above the header, it disappears when scrolling down (similar to announcement bar)
2. It will only shift the content the first time the page is loaded
3. Every subsequent load does not incur content shift
3. The "latest" tag is configurable via `mkdocs.yml`
4. The content can be freely defined (similar to announcement bar)

It doesn't use any translations, which means the author has to use theme extension to define it, but given the many different ways that those banners may look, I think it's the best approach. In the end, it boils down to defining a single template block:

``` html
{% block outdated %}
  You're not viewing the latest version of PRODUCT X.
  <a href="{{ config.site_url }}/{{ config.extra.version.latest }}">
    Click here to go to latest.
  </a>
{% endblock %}
```

It would be perfect to fetch the latest version tag from `mike`, but since `mike` is run on-top of `mkdocs`, that's not possible. However, the latest version tag should rarely change, so this shouldn't be a problem.

Also, I've discontinued following the "stick to bottom" solution, because it would steal too much space and might introduce problems with the copy to clipboard dialog or the back-to-top button.

**Comment by squidfunk** [comment]
The [version warning](https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#version-warning) feature is now available!

![version-warning](https://user-images.githubusercontent.com/932156/112762710-6cd1fc00-9001-11eb-8fce-8324dd65db81.png)

**Reply by JulioV** [reply]
thanks a lot!

**Reply by miguno** [reply]
This looks great, thank you!
